server/plugin.go

```diff
@@ -410,26 +410,27 @@ func (p *Plugin) restart() {
 
 func (p *Plugin) generatePluginSecrets() error {
 	needSaveConfig := false
-	if p.configuration.WebhookSecret == "" {
+	cfg := p.getConfiguration()
+	if cfg.WebhookSecret == "" {
 		secret, err := generateSecret()
 		if err != nil {
 			return err
 		}
 
-		p.configuration.WebhookSecret = secret
+		cfg.WebhookSecret = secret
 		needSaveConfig = true
 	}
-	if p.configuration.EncryptionKey == "" {
+	if cfg.EncryptionKey == "" {
 		secret, err := generateSecret()
 		if err != nil {
 			return err
 		}
 
-		p.configuration.EncryptionKey = secret
+		cfg.EncryptionKey = secret
 		needSaveConfig = true
 	}
 	if needSaveConfig {
-		configMap, err := p.configuration.ToMap()
+		configMap, err := cfg.ToMap()
```

Comment:
With the `Clone` above, we also have to call `p.setConfiguration(cfg)` if we intend to rely on the configuration change immediately. But then that causes a race in that we might be clobbering another update to the config that happened synchronously.

I wonder if we could replicate what Playbooks had to do here? See https://github.com/mattermost/mattermost-plugin-playbooks/blob/master/server/config/service.go, but in short configuration updates are done under a lock with a callback to effectively handle the direct patching and server update.

Reply:
this seems to be already the case. where it differs is that we call this function `generatePluginSecrets` when activating the plugin, and if needed it will call the API SavePluginConfig method and rely on the `onConfigurationChange` hook.

So what I'm doing now is cloning the config and also calling `setConfiguration` so that is set locally before calling `SavePluginConfiguration` on the API

I think this should take care of the problem